FirebaseDatabaseUI/FirebaseIndexArray.m

```diff
@@ -0,0 +1,205 @@
+// clang-format off
+
+//
+//  Copyright (c) 2016 Google Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+// clang-format on
+
+#import "FirebaseIndexArray.h"
+#import "FirebaseQueryObserver.h"
+
+@interface FirebaseIndexArray () <FirebaseArrayDelegate>
+
+@property (nonatomic, readonly) id<FIRDataObservable> index;
+@property (nonatomic, readonly) id<FIRDataObservable> data;
+
+@property (nonatomic, readonly) FirebaseArray *indexArray;
+
+@property (nonatomic, readonly) NSMutableArray<FirebaseQueryObserver *> *observers;
+
+@end
+
+@implementation FirebaseIndexArray
+
+- (instancetype)init {
+  NSException *e =
+    [NSException exceptionWithName:@"FIRUnavailableMethodException"
+                            reason:@"-init is unavailable. Please use the designated initializer instead."
+                          userInfo:nil];
+  @throw e;
+}
+
+- (instancetype)initWithIndex:(id<FIRDataObservable>)index
+                         data:(id<FIRDataObservable>)data
+                     delegate:(nullable id<FirebaseIndexArrayDelegate>)delegate; {
+  NSParameterAssert(index != nil);
```

Comment:
I tried this and ended up duplicating much of the `index` logic since queries only give you the `previousChildKey`. Might be worth extracting that from FirebaseArray into a more inert data object, but for now I think keeping FirebaseArray here is cleaner, if not less verbose.